re-implement pickValidInfo dataDir, move to quorum calculation

harshavardhana opened the pull request:
## Description
re-implement pickValidInfo dataDir, move to quorum calculation

## Motivation and Context
dataDir loosely based on maxima is incorrect and does not
work in all situations such as disks in the following order
    
- xl.json migration to xl.meta there may be partial xl.json's
  leftover if some disks are not yet connected when disk
  is yet to come up, since xl.json mtime and xl.meta is
  same the dataDir maxima don't work properly leading to
  quorum issues.
    
- its also possible that XLV1 might be true among the disks 
  available, make sure to keep FileInfo based on common quorum
  and skip unexpected disks with the older data format.

## How to test this PR?
CI/CD will do the job for you 🍰 

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation updated
- [x] Unit tests added/updated

harshavardhana committed:
add upgrade github workflow

tests upgrade from older to a newer release if
the data is readable and matches the checksum.

NOTE: this is just initial work we can build
on top of this to do further tests.

harshavardhana committed:
reduce number of workflows for go1.16

harshavardhana committed:
fix: upgrade tests for older releases

donatello commented on .github/workflows/upgrade-ci-cd.yaml:
```diff
@@ -0,0 +1,74 @@
+          docker-compose -f "buildscripts/upgrade-tests/compose.yml" up -d --build
+          until (mc alias set minio http://127.0.0.1:9000 minioadmin minioadmin) do echo "...waiting..." && sleep 5; done;
+          mc mb minio/minio-test/;
+          mc cp /usr/bin/emacs minio/minio-test/to-read/;
```
Lol, I like this usage of emacs!

harshavardhana committed:
re-implement pickValidInfo dataDir, move to quorum calculation

dataDir loosely based on maxima is incorrect and does not
work in all situations such as disks in following order

- xl.json migration to xl.meta there may be partial xl.json's
  left over if some disks are not yet connected when disk
  is yet to come up, since xl.json mtime and xl.meta is
  same the dataDir maxima doesn't work properly leading to
  quorum issues.

- its also possible that XLV1 might be true among the disks
  available, make sure to keep FileInfo based on common quorum
  and skip unexpected disks with older data format.
